source/adapters/cuda/device.cpp

```diff
@@ -809,9 +807,21 @@ UR_APIEXPORT ur_result_t UR_APICALL urDeviceGetInfo(ur_device_handle_t hDevice,
   case UR_DEVICE_INFO_GLOBAL_MEM_FREE: {
     size_t FreeMemory = 0;
     size_t TotalMemory = 0;
+    // This driver api call requires a CUcontext set on the device. In the case
+    // that a CUcontext is not already initialized on the device, we implicitly
+    // choose to momentarily initialize the primary CUcontext of the device (by
+    // retaining it). The CUcontext is immediately released in case the device
+    // is not later used. In the unlikely circumstance that the primary
+    // CUcontext has not already been retained by the programmer, the implicit
+    // initialization/destruction here adds a ~0.1s overhead, but this ensures
+    // SYCL specification compliance and MPI compatibility.
+    CUcontext tmpContext;
+    UR_CHECK_ERROR(cuDevicePrimaryCtxRetain(&tmpContext, hDevice->get()));
+    ScopedContext Active(tmpContext);
     detail::ur::assertion(cuMemGetInfo(&FreeMemory, &TotalMemory) ==
                               CUDA_SUCCESS,
                           "failed cuMemGetInfo() API.");
+    UR_CHECK_ERROR(cuDevicePrimaryCtxRelease(hDevice->get()));
     return ReturnValue(FreeMemory);
```

Comment:
I seem to remember there was multiple properties that needed a context, are you sure this is enough?

Reply:
Yeah I'm pretty sure:

sycl-ls tests all supported aspects by the test device that are exposed in dpc++. This includes examples of

- `cuDeviceGetPCIBusId`
- `cuDeviceGetUuid_v2` or `cuDeviceGetUuid` depending on toolkit version
- `cuDriverGetVersion`
- `cuDeviceGetAttribute` : this is most cases
- `cuDeviceGetName`
- `cuDeviceGetPCIBusId`

`cuDeviceTotalMem` is tested in another test without creating a context explicitly or implicitly: https://github.com/intel/llvm/blob/5cd9de100f8df3692b492e22d056e88798873ceb/sycl/test-e2e/syclcompat/device/device.cpp#L167

This leaves  `cuMemGetInfo` as the only one that fails without `CUcontext` set from the sycl-ls test. This is the only api that takes a variable that depends on a CUcontext being set: I think the api tells you how much memory is left for the CUcontext to use.